### DISCUSSION TITLE
Revisit rules for OAuth usage in config flow (for Google)

### DISCUSSION THREAD

**Discussion by allenporter** [home-assistant/architecture#692] [General]
## Summary

I'd like to make setup/onboarding easier for Nest and Google integrations by revisiting our design standards for OAuth

## Current State
Our current [stance](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md) for integrations that expose devices is that config flows should be used instead of configuration yaml. However there is some nuance for OAuth which I interpret as something like:

> - We prefer account linking. Nabu Casa works with the integration owner / cloud provider on Cloud Account Linking to set this up.
> - Otherwise, OAuth client id and secret stay in `configuration.yaml` since this is an advanced feature

## Proposal
Change our policy to be:
> - We prefer account linking. Nabu Casa works with the integration owner / cloud provider on Cloud Account Linking to set this up.
>    - **If a workable solution is not possible, grant an exception to allow Client ID / Secret to be used in the config flow**
> - Otherwise, OAuth client id and secret stay in `configuration.yaml` since this is an advanced feature

In the last year since proposing https://github.com/home-assistant/core/pull/44529 I've worked with Nabu Casa and Google on assessing feasibility of an account linking process for `nest`. The current state is that there is not a workable solution.

I am proposing we allow for Google and other integrations what we currently allow for Tuya. 
 Specifically for Google integrations, the work involved is:
- Move Nest (and all other Google integrations) to use Installed App Auth flow.
    - [Done] This was completed for Nest in https://github.com/home-assistant/core/pull/59033 (also has some discussion about calendar)
- Use auth setup that is compatible across Google integrations. 
- Allow Config Flow w/ Client ID & Secret with beefed up step-by-step walkthroughs

## Additional Notes
- My 2021 review of [Home Assistant + Google Authentication](https://docs.google.com/document/d/1w7ZxUvsFaWvKuAoTLUltlJUwyR3NenFlQB1TMkiiYBI/edit#) that i've been using to revamp nest authentication and config flow. Has notes from a discussion with @balloob where we sketched out this proposal.

- I surveyed the top ~50 integrations that involve auth, and figured it was worth sharing this summary:
  - OAuth:
    - Google Assistant, Spotify, Alexa, Netatmo, Xbox all use Cloud Link for OAuth
    - Google Calendar, and Nest use Client ID & Secret in yaml. (Google Assistant also has an option for a service account for pushing data which works differently.)
    - Tuya accepts Access ID, Access Secret but also Username, Password in config flow
  - Other:
    - Smart Things uses a personal access token however [OAuth seems like a possible option](https://developer-preview.smartthings.com/docs/connected-services/oauth-integrations)
    - Xiaomi Miio uses usernames and passwords but might [support oauth](http://docs.api.xiaomi.com/en/talos/api/outer-auth.html)

- My prior 2020 review of [Home Assistant Nest Config Flow](https://docs.google.com/document/d/1uhxJZz2G2hqNwJCNH7bblqqPKD5kxrj0n2u7EUQgITQ/edit#)
- We could have a shared place in the UI in home assistant to enter client ID and secret could be possible (e.g. somewhere else in settings), but does not change what the user has to do (enter a client id and secret). However, it seems simplest is to just let users share  the client/id secret between the two projects.  We could alternatively have a shared authentication library somewhere for Google creds if helpful, but not sure it is worth the complexity.

**Comment by MartinHjelmare** [comment]
I'm not sure how much more easy it'll be if we allow client id and secret in the config flow. The hard part is the cloud application setup, not pasting the client credentials in configuration.yaml or in the config flow.

I think it's a good idea to move to installed app auth flow for Google. It's working well in my WIP branch for Google calendar.

**Reply by allenporter** [reply]
These are my thoughts when it comes to ease of use:
- Last week: I saw https://community.home-assistant.io/t/set-up-nest/365053 
   > i am trying to setup nest in HA, i have followed all steps and think i’ve done it right but when i try to install nest it just comes up saying Aborted The component is not configured. Please follow the documentation.
- On `#integrations` on discord today:
    > I followed all of the device access registration steps, completed the configuration.yaml with the correct ID's, restarted Home Assistant, and then when I try to add the integration, I receive error "Aborted. The component is not configured. Please follow the documentation." 
    - When I told them their configuration.yaml was invalid they disagreed with me and inisted it was correct.
- Using a config flow we could have a step by step interactive set of instructions corresponding to the 4 steps in the [Device Access Registration](https://www.home-assistant.io/integrations/nest/).
- Supporting some kind of device discovery would be nice, but it seems like a no go to involve `configuration.yaml` in the middle of that flow.

Thanks for the discussion.  Setup has gotten much simpler for users with the move to installed app auth and automated pubsub creation.

**Reply by balloob** [reply]
I think that it will become easier because at least we don't steer users to configuration.yaml.

**Reply by allenporter** [reply]
I've added tests in https://github.com/home-assistant/core/pull/62526 that show how nest discovery is currently *effectively* unsupported due to a config flow based approach not being allowed, and requiring configuration.yaml.

**Reply by allenporter** [reply]
I'd like to revive the discussion on usability.  I have implemented this yet again in https://github.com/home-assistant/core/pull/64680 and https://github.com/home-assistant/home-assistant.io/pull/21303 so we have something concrete we can compare with.

While this has definitely not removed the complexity of the nest setup by any means, I've verfied the usability improvements above:
- Discovery now works, and can find nest devices on the network
- Less confusing error message when `configuration.yaml` is not present (abort status `missing_configuration`) when adding the integration, and instead jumps right to setup with instructions.
- The interactive config flow has markdown that links exactly where you need to go and mirrors 1:1 the existing documentation steps
- One  trip to cloud console to get the cloud project id was removed. This step often confuses users as its one more chance to get "Cloud Project ID" and "Device Access Project ID" confused
- No more configuration.yaml, no more restarts required.
- Additional back and forth steps are removed (e.g. "write this down in a text file").  Additional steps that required restarts are removed.

So, none of this is hard, but seems pretty clearly aligned with making home assistant easier to use. That said, you are right that all the improvements made in the previous year around Installed App Auth and streamlined documentation have helped significantly.

Any other thoughts with regard to usability?

**Comment by balloob** [comment]
I think this is a good idea. I wonder if we should adjust the exception to be for cases where the client ID/secret are bound to a single user account? (which is the case here for Nest)

**Reply by allenporter** [reply]
Yes, that is true, it needs to be for the account owner.  )Are there cases where people give instructions to use a shared client id and secret? oh man...)

**Comment by frenck** [comment]
I would opt for a separate page/panel to add developer accounts (or name it something similar) to register OAuth implementation and NOT make it part of the configuration flow.

What we can do in the config flow, is detect there are no implementations -> redirect the user to the flow to add/register an implementation/developer account. After adding such an account, we can restart the integration setup.

The idea would be similar to blueprints. Register developer accounts/connections to use with integration instances.

It keeps them decoupled, re-useable, and manageable.

**Reply by balloob** [reply]
I like this idea too. That way it could also work for any other integration and no exceptions are needed.

**Reply by allenporter** [reply]
When I start to think through the implementation of this, I consider that we have:
- authentication provider specific details (e.g. the redirect url, the authentication library to use, etc). Often these are 1:1 to the integration itself. 
- integration specific details: This might include variations on how the auth provider is configured (e.g. in the case of `nest` and `google` they have different authorization urls).  They'd also integrate into the config flow in different ways to collect integration specific details
- Account specific configuration (e.g. oauth client id and secret)

So, when considering the relationships between these things, i consider these cases:
1. an Integration and dev accounts with a 1:1 mapping and a single config entry
1. an integrations with multiple dev accounts:  one config entry for each account
1. dev accounts shared across multiple integrations 

- I'm assuming the first case is the most popular and should be the primary use case.
- The second case seems like it could make it harder to support easy config flow integration (e.g. how do you know if the user wants to proceed with integration setup vs add another account? Maybe there is an "Add account" page where you can manually add more accounts when outside the config flow
- The third case.. are there others besides? `nest` and `google`. I haven't thought through how to structure this, but maybe there is a way to make one integration the "primary" domain to pull from, or just punt on this.


If we focus on the first two cases:  Would modeling an "AccountFlow" flow after an "OptionsFlow" be close?

**Reply by allenporter** [reply]
Before proceeding with implementing this as a config flow for nest, I wanted to follow up on the idea of a separate developer account page. I accidentally replied to the wrong thread below with my survey of integrations using OAuth.

What integrations would be a candidate for decoupled auth vs integration instance?

I am happy to push this forward if there are use cases that can use it -- or, at a minimum could implement things moving towards this long term direction.  (e.g. decouple parts of the config flow, or extend the OAuth API abstractions to call out to a separate part of the config flow to handle this).

**Reply by allenporter** [reply]
I'd like to propose that we don't yet have use cases for proceeding with this direction, pending some concrete use cases. I've done the data collection here and below and haven't been able to find them in order to inform what a design would like, but recognize there may be use cases that I have not considered (e.g. re-using oauth credentials for multiple config entries in a single integration, but i haven't been able to find cloud APIs that work that way yet)

**Reply by MartinHjelmare** [reply]
Fitbit works like that.

**Reply by allenporter** [reply]
I may be also misunderstanding the point.  I think the use case for using credentials across integrations isn't there, but is it more about having a streamlined way for integrations to accept oauth credentails?

Having a streamlined way to accept oauth credentials for simplifying the code base  or providing consistency across integrations for users could be interesting.  Te 1:many case is what i'm skeptical about and interested in understanding the concrete use cases.

**Reply by frenck** [reply]
Reusable application credential for integrations, from the top of my head, this is possible for: Spotify, Toon, FitBit, (some) google integrations, Somfy, Netatmo, Withings, xbox.

Upcoming: Geocaching, Twitch, LaMetric 

Possibly, but unsure: Smappee, Neato, Lyric

Possible, but not right now for vendor reasons: Home Connect, Tuya

The above is from the top of my head, so I'm sure I missed a bunch.

**Reply by allenporter** [reply]
I see -- so the huge context I was missing is you want to use this for integrations that already have a well lit path for cloud linking?

**Reply by frenck** [reply]
> I was missing is you want to use this for integrations that already have a well lit path for cloud linking?

Yes, a global feature in handling application/developer credentials.

**Comment by allenporter** [comment]
One reason I did the survey of top integrations is to see what kind of usage something like this would get. Looking at the top 100 integrations I believe these 4 are eligible:
- Tuya (13% of installs) Currently accepting client id and secret in the config flow, and it has a one-page signup with country, access id, secret, username, and password.
- Google (Calendar) (6% of installs) - Uses configuration.yaml 
- Nest (4% of installs) - Uses configuration.yaml
- Home Connect  (1% of installs) - Uses configuration.yaml

Has there been discussion with Tuya or Home Connect about cloud linking? Basically, this is why I propose just using the config flow is that i'm not sure there would be sufficient use for a central user interface for this.

How I interpret this variation on the proposal is that we actually do not have a problem with users entering client id and secret into text boxes when setting up integrations. Perhaps we could take a step back away from implementation approach and discuss the principles with respect to OAuth?  I think i'm trying to reverse engineer these a bit, without understanding the user problem we're trying to solve.

**Reply by balloob** [reply]
Tuya is out of the question for account linking as they charge money for that. 

Home Connect we've had discussions but we kinda dropped the ball there.

**Reply by balloob** [reply]
The reason we implemented this policy is to enforce the ideal flow when account linking is available. If we allow asking client ID/secret, we should offer that option to the user even if account linking is available, causing confusion.

This proposal is about adding an exception in case a workable solution doesn't exist. I think that's a good idea ([also commented that above](https://github.com/home-assistant/architecture/discussions/692#discussioncomment-1842477))

**Reply by allenporter** [reply]
Yes, i thought i had some alignment that it made sense, but I was informed in the implementation PR that isn't the case. The status quo isn't acceptable, so i'm motivated to figure out a solution (even if its for satisfying less important use cases, IMO, given account linking is already available for those) and will revisit this after thinking about it more.

**Comment by allenporter** [comment]
Starting a side thread to go into a detailed overview of a global feature for developer credentials as a path forward to solving the Nest use case which has no options for any streamlined paths today. (Other integrations already have a streamlined path for cloud link, but have a broader need for infrastructure simplicity in order to support much more complex configuration use cases that cannot be supported with yaml)

### Developer Credentials 

A new component will be added for managing **Developer Credentials**. A Developer Credential is:
- client id
- client secret
- an identifier e.g. (domain, client id)

The credentials component can support these operations (e.g. websocket/storage) similar to blueprints:
- list: Return all developer accounts (domain, client id)
- add: Create developer credential identified by (domain, client id) with the client secret
- delete: By identifier. A developer credential may not be deleted if it is currently in use by a config entry.

Users have an expectation that a deleting a credential means it is completely gone (all copies are gone, all access tokens or refresh tokens are gone, etc). Edit is not a supported operation.

### Example user flow end to end

As an example user journey for illustration:
- Integration registers that it supports Developer Credentials
- User starts config flow. If the integration supports cloud link, that is still used by default with no change. Otherwise, user is instructed to register a developer credential, or is automatically navigated there. (compared with a `missing_configuration` error today)
- User creates a Developer Credential. (Note: User may also start at this step)
- User resumes the config flow. If there is just one auth implementation the flow would proceed as normal. Otherwise, it acts just like auth implementations do today.

An integrations config flow remains unchanged, and relies on the oauth config flow to pick/manage/show auth implementations and create config entries on its behalf. The integration config flow doesn't need to know about details of credentials.

### Authorization Server and Authentication Impl

You can think of this effort like splitting up the responsibilities of a LocalOAuthImplementation, and changing up the registration mechanism.  Integrations are still responsible for declaring the **Authorization Server** to use: an authorize url, and token url. 

We consider these degrees of freedom between Developer Credentials, Authorization Server, and Config Entries to inform how registration works:
- Multiple Developer Credentials for an Authorization Server. *Use case: config entry for each account that I have.* 
- Multiple Authorization Servers for an integration. *Use case: Different auth types e.g. Installed App auth, Web Auth, backwards compatibility needed if preferred types change over time*
- Multiple uses of the same Developer Credential and Authorization Server: *Use case: separate config entry per cloud device, in above fitbit example.*
- Authorization Server config varies per Config Entry. *Use case: An authorization url contains a nest project id*

The number of combinations implies to me we may want a pretty flat representation of these things, but we can get into this in detailed design next.

Integration setup will register an Authorization server, and Developer Credentials for any accounts currently in yaml.  Integration setup for a config entry is similar to today (e.g. asks for an auth implementation for a given config entry). The config entry has a pointer/reference to a developer credential, which is also used for delete protection above.

When considering the user experience, integrations may need to customize Developer Credentials setup with extra instructions/markdown or links -- though many integrations with trivial authentication setup may not even need this.

A handful of integrations will operate on Developer Credentials at a lower level, similar to how yaml compatibility works, or how oauth config flows are customized today. This keeps integration specific complexity within the integrations, allowing for more integrated user experience while keeping the developer credentials APIs simple e.g. ([screenshot #1](https://user-images.githubusercontent.com/6026418/150669756-f01efa99-69fd-4d39-96fd-28849715a0a2.png), [screenshot #2](https://user-images.githubusercontent.com/6026418/150669759-6bd751b6-db94-4621-9f9a-762696819b96.png)) shows an example config flow that constructs inputs that appear in the Authorization Server URL, with an integrated setup to minimize bounding back and forth between consoles decreasing likelihood of mismatched credentials. Or put another way, custom config flows *must* use the developer credentials infrastructure and wiring.

I would like feedback on this high level direction before getting into the next level of implementation details.

**Reply by balloob** [reply]
I think this sounds good. Will be a great user experience.

For implementation we can leverage a `developer_credentials.py`platform to allow integrations to indicate they support developer credentials and provide the authentication implementation.

We can use the WebSocketCollection helper (see any input_X integration) to quickly stamp out a collection that supports create/read/update/delete.

**Reply by allenporter** [reply]
Continuing on with the API design at another level of detail, there are two relevant data classes to consider:
```
@dataclass
class AuthorizationServer:
    authorize_url: str
    token_url: str

@dataclass
class DeveloperCredential:
    client_id: str
    client_secret: str
```
An integration would replace `LocalOAuth2Implementation` registration with `a developer_credentials.py` platform that implements this protocol:

```
  async def async_get_authorization_server(
      self, hass: HomeAssistant
  ) -> AuthorizationServer:
      """Return authorization server."""

  async def async_get_developer_credential(
      self, hass: HomeAssistant
  ) -> DeveloperCredential:
      """Return a developer credential from configuration.yaml."""
```

The `async_get_developer_credential` method is optional, and only needed when yaml based configuration is specified. We could theoretically represent more than `DeveloperCredential` this way, however my impression is that all existing uses typically only have one credential at the moment, so we'll only support one.


The web socket create schema will allow creation of additional credentials for any particular integration domain:
```
CREATE_FIELDS = {
    vol.Required(CONF_DOMAIN): cv.string,
    vol.Required(CONF_CLIENT_ID): cv.string,
    vol.Required(CONF_CLIENT_SECRET): cv.string,
}
```
Question: I think we may also need a websocket to return all registered integrations with a `developer_credentials` platform.   Can we load platforms for an integration before the integration itself has been loaded?  e.g. we want to use the websocket to create a developer credential for a specific integration supporting this platform, before actually adding that integration.

Each `DeveloeprCredential` created, either via the websocket or via the integration platform, will correspond with a new auth implementation that is registered with `config_entry_oauth2_flow.async_register_implementation`.  The implementation will be registered under the integration domain and the auth domain will be unique to the (domain, client id). (Note: This id will eventually be persisted in the config entry `auth_implementation`)  The OAuth config flow registration method will be updated to return an unsubscribe method that will unregister the auth implementation when it is deleted.

We will also need a low level API used by integrations that need to dynamically create an `AuthorizationServer` during a config flow like Nest.  The config flow use case will effectively need to be able (1) to create a DeveloperCredential (e.g. like the websocket) and (2) register an AuthorizationServer to use with the specific credential with an API like:
```
async def async_register_auth_implementation(
    hass: HomeAssistant,
    domain: str,
    authorization_server: AuthorizationServer,
    auth_domain: str,
) -> None:
```
This same API will be used on startup to register the auth implementation from the config entry.

I imagine the following PRs would be needed:
- Implement the websocket and storage collection
- Add the platform, and register auth implementations for developer credentials
- Support for DeveloperCredentials in the protocol to handle existing YAML cases
- Support for manual registration during config flow (e.g. for dynamic authorization server urls for nest)
- Implement frontend

Edit: I realize I also haven't explained the config flow integration to redirect to create developer credentials yet. I think it may need additional methods exposed in the config flow oauth helper that indicates the domain is registered with developer credentials and returns a different return code to the frontend to setup developer credentials.

**Reply by balloob** [reply]
Ooof, a month. Sorry for the late reply.

We can either add a manifest field that it allows custom developer credentials, or we check for `components/*/developer_credentials.py` files and list them with hassfest in `generated/developer_credential.py`.

For authorizaton URL, should we also include redirect_uri or are we still assuming that the user will want to redirect to the URL that they are currently visiting? Especially if we want to offer users without HTTPS an HTTPS proxy endpoint to use.

**Reply by allenporter** [reply]
Ok, I'm not familiar with what the hassfest part will do but will look into it.

My impression was that this is a authorize base url and the config flow will add the redirect url and state and other things to it.  Interface is the same as local oauth impl. (So redirect url is already handled)

---

Are we broadly aligned? @frenck 

In particular, the bit about still collecting creds in a config flow UI, but using developer creds storage, I expect is the most controversial part (a variation on a previous PR I sent with different storage). Given nest already has two types of developer creds, and I will likely add another given the recent deprecation, I don't think we want to spread that complexity of creds types into the interface for this as it seems specific to google. This is my proposed compromise to help push forward vision of developer creds, which can be orthogonal (IMO) from a UI perspective.

Last time I had the mistaken impression we had alignment and do not want to waste people's time sending PRs.  Initial core PRs for this integration proposed have been ready to go for about a month or so, or happy to incorporate feedback.

**Reply by allenporter** [reply]
With respect to the manifest field or codifying via hassfest and a generated file:  I generally am all behind the value of codifying intent in a manifest, though don't have a clear understanding of the practical implications.  Can you elaborate on the motivations behind this?   I was comparing to diagnostic, which I didn't think had this.

**Reply by balloob** [reply]
We prefer platforms if there is code related to it, or else there is not a single source of truth. Config flow would be a good candidate to remove from the manifest as it can be derived from existence of `config_flow.py`

**Reply by raman325** [reply]
I have truthfully not read all of the comments in this thread, which I will do later and edit this post as needed, but I love this idea! It never made sense to me that users were forced to use both configuration.yaml and config flow for setup of an oauth integration, so I am all in favor of bring it into the UI, and creating a dedicated platform for this makes sense to me.

Two things to think about:
1. User attempts to delete credentials for existing config entries, what happens? IMO we shouldn't block this action, we should do what we can to help them. So my suggestion is to tell the user, hey, xyz config entries are linked to this credentials and will be deleted along with the credentials. Do you want to proceed?
2. User deletes a config entry leaving a set of oauth credentials that are no longer tied to a config entry, what happens? The default and easiest solution is to let them go to this new dashboard and delete the credentials after, but we could add a confirmation step to the removal process that allows the user to delete the credentialsl as part of the config entry removal

**Reply by allenporter** [reply]
> We prefer platforms if there is code related to it, or else there is not a single source of truth. Config flow would be a good candidate to remove from the manifest as it can be derived from existence of `config_flow.py`

I get why we want a playform -- i just don't see what `generated/developer_credential.py` is needed yet. Perhaps because I haven't added the websocket call that needs to list all supported platforms yet.

**Reply by allenporter** [reply]
> I have truthfully not read all of the comments in this thread, which I will do later and edit this post as needed, but I love this idea! It never made sense to me that users were forced to use both configuration.yaml and config flow for setup of an oauth integration, so I am all in favor of bring it into the UI, and creating a dedicated platform for this makes sense to me.
> 
> Two things to think about:
> 
> 1. User attempts to delete credentials for existing config entries, what happens? IMO we shouldn't block this action, we should do what we can to help them. So my suggestion is to tell the user, hey, xyz config entries are linked to this credentials and will be deleted along with the credentials. Do you want to proceed?
> 2. User deletes a config entry leaving a set of oauth credentials that are no longer tied to a config entry, what happens? The default and easiest solution is to let them go to this new dashboard and delete the credentials after, but we could add a confirmation step to the removal process that allows the user to delete the credentialsl as part of the config entry removal

Thank you, these are great suggestions.

**Reply by frenck** [reply]
Read the above, the generic flow of things matches with what I had in mind, so from my end 👍 
The idea to add a platform for it is great btw, there are probably tons of little details we run into, but the big picture sounds like the right path to take.

**Reply by frenck** [reply]
> Config flow would be a good candidate to remove from the manifest as it can be derived from existence of config_flow.py

Yes/no. Not sure if it is still the case, but we had a couple of cases that used an import flow, but not a user-based flow. In that case, there would be a config flow, but not marked as such in the manifest (so it would not show up in the UI).

**Reply by frenck** [reply]
So, looking at upstream implementations, my biggest concern is the simplest one: "Developer Credentials"

I think that name is incorrect and not really end-user-friendly. The term "developer" scares people unneeded. I would suggest "Application Credentials", or (as specs call it) "Client Credentials"

**Reply by allenporter** [reply]
Agreed, I had never heard the term "developer" used in this context before until this discussion thread.

(1) For developer end users, I do think it makes sense to be aligned to the spec e.g. "Client Credentials" so happy to go with that.

(2) For actual home assistant end users, my impression is we would pick a name for the interface that makes sense in that context and it would be different from the name used in the component.  e.g. w/ "Applications", "Accounts", "Linked" in the name

**Reply by allenporter** [reply]
For (1) I'm warming up to using Application Credentials as the component name, and then Client Credentials for the client id and secret.

**Comment by frenck** [comment]
Closing this architectural topic, as we have covered this now.

../Frenck
